File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/validation/NotStartsWithValidator.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.validation;

import java.util.Set;

import javax.validation.constraints.NotNull;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

/**
 * A Validator that asserts that the attribute value does
 * <b>NOT START WITH</b> any of the invalid prefixes.
 */
@AllArgsConstructor
@Getter
@Setter
public final class NotStartsWithValidator implements SerializableAttributeValidator {

    private static final long serialVersionUID = 1L;

    @NotNull
    private Set<String> invalidPrefixes;

    /**
     * Validates that the attribute value does not start with any of the invalid
     * prefixes.
     */
    @Override
    public boolean eval(String value) {
        if (value == null) {
            return false;
        }

        return invalidPrefixes.stream().noneMatch(prefix -> value.trim().startsWith(prefix));
    }

}

File: lib/python/opentoken/src/main/opentoken/attributes/validation/not_starts_with_validator.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import Set
from opentoken.attributes.validation.serializable_attribute_validator import SerializableAttributeValidator


class NotStartsWithValidator(SerializableAttributeValidator):
    """
    A Validator that asserts that the attribute value does NOT START WITH
    any of the invalid prefixes.
    """

    def __init__(self, invalid_prefixes: Set[str]):
        """
        Initialize the validator with a set of invalid prefixes.

        Args:
            invalid_prefixes: Set of prefixes that values should not start with

        Raises:
            ValueError: If invalid_prefixes is None
        """
        if invalid_prefixes is None:
            raise ValueError("Invalid prefixes set cannot be None")

        self._invalid_prefixes = invalid_prefixes.copy()

    @property
    def invalid_prefixes(self) -> Set[str]:
        """Get the set of invalid prefixes."""
        return self._invalid_prefixes.copy()

    def eval(self, value: str) -> bool:
        """
        Validate that the attribute value does not start with any of the invalid prefixes.

        Args:
            value: The attribute value to validate

        Returns:
            True if the value doesn't start with any invalid prefix, False otherwise
        """
        if value is None:
            return False

        trimmed_value = value.strip()
        return not any(trimmed_value.startswith(prefix) for prefix in self._invalid_prefixes)
